services/src/main/java/org/keycloak/services/clientpolicy/DefaultClientPolicyManager.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.clientpolicy;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.Profile;
+import org.keycloak.component.ComponentModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.services.clientpolicy.ClientPolicyContext;
+import org.keycloak.services.clientpolicy.ClientPolicyException;
+import org.keycloak.services.clientpolicy.ClientPolicyManager;
+import org.keycloak.services.clientpolicy.ClientPolicyProvider;
+import org.keycloak.services.clientpolicy.condition.ClientPolicyConditionProvider;
+import org.keycloak.services.clientpolicy.executor.ClientPolicyExecutorProvider;
+
+public class DefaultClientPolicyManager implements ClientPolicyManager {
+
+    private static final Logger logger = Logger.getLogger(DefaultClientPolicyManager.class);
+
+    private final KeycloakSession session;
+    private final Map<String, List<ClientPolicyProvider>> providersMap = new HashMap<>();
+
+    public DefaultClientPolicyManager(KeycloakSession session) {
+        this.session = session;
+    }
+
+    @Override
+    public void triggerOnEvent(ClientPolicyContext context) throws ClientPolicyException {
+        if (!Profile.isFeatureEnabled(Profile.Feature.CLIENT_POLICIES)) return;
```

Comment:
@tnorimat @stianst I have some concerns about the security of this.

I am thinking of some scenario like:
- Administrator start the server with the CLIENT_POLICIES feature enabled
- Administrator configures some client policies
- Administrator restart the server, but he forgets to enable the CLIENT_POLICIES feature for some reason (For example he is migrating from one environment to another and startup commands are slightly different etc).
- Client policies now won't be triggered at all, which can result in some potential unexpected issues (in theory even security issues).

Do we want some better protection against this scenario? For example if CLIENT_POLICIES feature is not enabled, BUT there are some client policies in the DB, then throw the exception at startup - or at least big warning.

I've just checked that for example WebAuthn works this way (If there is WebAuthn feature disabled BUT WebAuthn required action is registered in the realm, then the exception is thrown during attempt to authenticate user).

Reply:
@mposolda  Right now the client policies are preview, so the reason why there's a profile guard. I think when we mark it as supported we'd just remove it from the profile and have it always on. Since it's preview some great care should be considered before using it in production. With that in mind I don't think we need to introduce any extra guards, and can consider that in the future when we mark it as supported if we want to keep it as a feature that can be enabled/disabled. Does that work for you?

Reply:
Yes, this is not the reason to block this PR in my opinion.

This is probably not specific to this PR, but rather a general thing. I am not sure if some better care should be taken for various features/profiles to be enabled, configured and then accidentally disabled after migration to another environment. I can see the potential for unexpected issues and some of them even security related. But maybe I am just overthinking this :) We can see if someone reports some similar issue in the future...